backends/tfhe-cuda-backend/cuda/src/crypto/keyswitch.cuh

```diff
@@ -146,12 +147,169 @@ void execute_keyswitch_async(cudaStream_t *streams, uint32_t *gpu_indexes,
         GET_VARIANT_ELEMENT(lwe_input_indexes, i);
 
     // Compute Keyswitch
-    cuda_keyswitch_lwe_ciphertext_vector<Torus>(
+    host_keyswitch_lwe_ciphertext_vector<Torus>(
         streams[i], gpu_indexes[i], current_lwe_array_out,
         current_lwe_output_indexes, current_lwe_array_in,
         current_lwe_input_indexes, ksks[i], lwe_dimension_in, lwe_dimension_out,
         base_log, level_count, num_samples_on_gpu);
   }
 }
 
+// chunk_count = glwe_size * polynomial_size / threads.
+// each threads will responsible to process only lwe_size times multiplication
+template <typename Torus>
+__device__ void keyswitch_lwe_ciphertext_into_glwe_ciphertext(
+    Torus *glwe_out, Torus *lwe_in, Torus *fp_ksk, uint32_t lwe_dimension_in,
+    uint32_t glwe_dimension, uint32_t polynomial_size, uint32_t base_log,
+    uint32_t level_count) {
+
+  extern __shared__ int8_t sharedmem[];
+
+  // result accumulator, shared memory is used because of frequent access
+  Torus *local_glwe_chunk = (Torus *)sharedmem;
+
+  const int tid = threadIdx.x + blockIdx.x * blockDim.x;
+
+  const int shmem_index = threadIdx.x;
+  // the output_glwe is split in chunks and each x-block takes one of them
+  size_t chunk_id = blockIdx.x;
+  size_t coef_per_block = blockDim.x;
+
+  // dimensions
+  size_t glwe_size = (glwe_dimension + 1);
+  // number of coefficients inside fp-ksk block for each lwe_input coefficient
+  size_t ksk_block_size = glwe_size * polynomial_size * level_count;
+
+  // initialize accumulator to 0
+  local_glwe_chunk[shmem_index] =
+      SEL(0, lwe_in[lwe_dimension_in], tid == glwe_dimension * polynomial_size);
+
+  // Iterate through all lwe elements
+  for (int i = 0; i < lwe_dimension_in; i++) {
+    // Round and prepare decomposition
+    Torus a_i = round_to_closest_multiple(lwe_in[i], base_log, level_count);
+
+    Torus state = a_i >> (sizeof(Torus) * 8 - base_log * level_count);
+    Torus mod_b_mask = (1ll << base_log) - 1ll;
+
+    // block of key for current lwe coefficient (cur_input_lwe[i])
+    auto ksk_block = &fp_ksk[i * ksk_block_size];
+    for (int j = 0; j < level_count; j++) {
+      auto ksk_glwe = &ksk_block[j * glwe_size * polynomial_size];
+      // Iterate through each level and multiply by the ksk piece
+      auto ksk_glwe_chunk = &ksk_glwe[chunk_id * coef_per_block];
+      Torus decomposed = decompose_one<Torus>(state, mod_b_mask, base_log);
+      local_glwe_chunk[shmem_index] -= decomposed * ksk_glwe_chunk[shmem_index];
+    }
+  }
+
+  // Persist
+  glwe_out[tid] = local_glwe_chunk[shmem_index];
+}
+
+// public functional packing keyswitch
+//
+// blockIdx.y - input
+// chunk_count = glwe_size * polynomial_size / threads.
```

Comment:
What is chunk_count?

Reply:
It's an idea inherited from the old packing keyswitch implementation. We launch `num_inputs * glwe_size * polynomial_size` threads and split them through multiple blocks.  Each block process a chunk of the coefficients.

I will rethink this naming and improve comments.